packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -0,0 +1,228 @@
+ * subscribers when events happen. Events will run asynchronously.
+ */
+export default class RedisEventBusService extends AbstractEventBusModuleService {
+  protected readonly config_: ConfigModule
+  protected readonly moduleOptions_: EventBusRedisModuleOptions
+  protected readonly moduleDeclaration_: ConfigurableModuleDeclaration
+  protected readonly logger_: Logger
+
+  protected queue_: Queue
+  protected shouldEnqueuerRun: boolean
+  protected enqueue_: Promise<void>
+
+  constructor(
+    { logger, configModule }: InjectedDependencies,
+    moduleOptions: EventBusRedisModuleOptions = {},
+    moduleDeclaration: ConfigurableModuleDeclaration,
+    singleton = true
+  ) {
+    // @ts-ignore
+    super(...arguments)
+
+    if (moduleDeclaration?.resources !== MODULE_RESOURCE_TYPE.SHARED) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "At the moment this module can only be used with shared resources"
+      )
+    }
+
+    this.moduleOptions_ = moduleOptions
+    this.moduleDeclaration_ = moduleDeclaration
+    this.config_ = configModule
+    this.logger_ = logger
+
+    if (singleton && moduleOptions.redisUrl) {
+      this.connect()
+    }
+  }
+
+  connect(): void {
+    // Required config
+    // See: https://github.com/OptimalBits/bull/blob/develop/CHANGELOG.md#breaking-changes
+    const connection = new Redis(this.moduleOptions_?.redisUrl!, {
+      maxRetriesPerRequest: null,
+    })
+
+    this.queue_ = new Queue(this.moduleOptions_.queueName ?? `events-queue`, {
+      connection,
+      prefix: this.moduleOptions_.queuePrefix ?? `${this.constructor.name}`,
+    })
+
+    // Register our worker to handle emit calls
+    new Worker(this.moduleOptions_.queueName ?? "events-queue", this.worker_, {
+      connection,
+      prefix: this.moduleOptions_.workerPrefix ?? `${this.constructor.name}`,
+    })
+  }
+
+  /**
+   * Calls all subscribers when an event occurs.
+   * @param {string} eventName - the name of the event to be process.
+   * @param data - the data to send to the subscriber.
+   * @param options - options to add the job with
+   * @return the job from our queue
+   */
+  async emit<T>(
+    eventName: string,
+    data: T,
+    options: Record<string, unknown> & EmitOptions = { attempts: 1 }
+  ): Promise<void> {
+    const opts: { removeOnComplete: boolean } & EmitOptions = {
+      removeOnComplete: true,
+      attempts: 1,
+    }
+    if (typeof options.attempts === "number") {
+      opts.attempts = options.attempts
+      if (isDefined(options.backoff)) {
+        opts.backoff = options.backoff
+      }
+    }
+    if (typeof options.delay === "number") {
+      opts.delay = options.delay
+    }
+
+    await this.queue_.add(eventName, { eventName, data }, opts)
+  }
+
+  /**
+   * Handles incoming jobs.
+   * @param job The job object
+   * @return resolves to the results of the subscriber calls.
+   */
+  worker_ = async <T>(job: BullJob<T>): Promise<unknown> => {
```

Comment:
thought (non-blocking): my immediate reflection on the logic here is that it should go in the core. I.e., the core EventBus would keep track of subscribers and call them correctly. And the different provider modules would just ensure that events are placed correctly in the underlying tech and that the EventBus is notified of new work as new jobs leave the queue.

This would allow retry logic like what we have here to work for any underlying tech and give us a nice facade to built improvements from.

That said, I am not sure of the technical implications of this change and think it is alright to keep it here for now.